File: csharp/Airforce/Resources/NotificationsResource.cs
using System.Text.Json.Nodes;

namespace Airforce;

/// <summary>Notification preferences, the in-app feed, and delivery-channel linking —
/// <c>/api/me/notification-prefs</c>, <c>/api/me/notifications</c>, <c>/api/me/channels</c>.
/// All endpoints require a session token.</summary>
public sealed class NotificationsResource : Resource
{
    internal NotificationsResource(Transport t) : base(t) { }

    public Task<JsonNode?> GetPrefsAsync(CancellationToken ct = default)
        => Transport.GetAsync("/api/me/notification-prefs", "session", null, ct);

    /// <summary>Partially update preferences: an absent field is unchanged;
    /// <c>quiet_hours: null</c> clears quiet hours. Unknown channel ids in
    /// <c>routing</c> are dropped server-side. Returns the updated preferences.</summary>
    public Task<JsonNode?> UpdatePrefsAsync(object patch, CancellationToken ct = default)
        => Transport.MethodAsync(HttpMethod.Patch, "/api/me/notification-prefs", "session", patch, ct);

    /// <summary>List feed items, newest first — <c>{items, unread}</c>. <paramref name="limit"/>
    /// is 1–100 (default 30); <paramref name="before"/> is the <c>created_at</c> cursor of
    /// the last item from the previous page.</summary>
    public Task<JsonNode?> ListAsync(int? limit = null, string? before = null, CancellationToken ct = default)
    {
        var query = new Dictionary<string, string?>
        {
            ["limit"] = limit?.ToString(),
            ["before"] = before,
        };
        return Transport.GetAsync("/api/me/notifications", "session", query, ct);
    }

    /// <summary>Mark specific feed items read. Returns <c>{updated, unread}</c>.</summary>
    public Task<JsonNode?> MarkReadAsync(IEnumerable<string> ids, CancellationToken ct = default)
        => Transport.PostAsync("/api/me/notifications/read", "session", new { ids }, ct);

    /// <summary>Mark every feed item read. Returns <c>{updated, unread}</c>.</summary>
    public Task<JsonNode?> MarkAllReadAsync(CancellationToken ct = default)
        => Transport.PostAsync("/api/me/notifications/read", "session", new { all = true }, ct);

    /// <summary>List linked delivery-channel identities plus the linkable channel ids —
    /// <c>{identities, available_channels}</c>.</summary>
    public Task<JsonNode?> ChannelsAsync(CancellationToken ct = default)
        => Transport.GetAsync("/api/me/channels", "session", null, ct);

    /// <summary>Start linking a delivery channel. The verification code is delivered
    /// through the channel itself and expires after 30 minutes. Bot channels accept an
    /// empty <paramref name="address"/> and answer with a one-time link code / deep link
    /// (<c>{status: "link_ready", code, deep_link?, expires_minutes}</c>) instead of
    /// <c>{status: "verification_sent"}</c>.</summary>
    public Task<JsonNode?> LinkChannelAsync(string channel, string address, string? display = null,
        CancellationToken ct = default)
    {
        object body = display != null ? new { channel, address, display } : new { channel, address };
        return Transport.PostAsync("/api/me/channels", "session", body, ct);
    }

    /// <summary>Complete channel verification with the delivered code (400 on an invalid
    /// or expired code).</summary>
    public Task<JsonNode?> VerifyChannelAsync(string channel, string code, CancellationToken ct = default)
        => Transport.PostAsync("/api/me/channels/verify", "session", new { channel, code }, ct);

    /// <summary>Unlink/revoke a channel identity.</summary>
    public Task<JsonNode?> UnlinkChannelAsync(string channel, CancellationToken ct = default)
        => Transport.DeleteAsync($"/api/me/channels/{Enc(channel)}", "session", ct);
}

File: java/src/main/java/com/airforce/NotificationsResource.java
package com.airforce;

import com.fasterxml.jackson.databind.JsonNode;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Notification preferences, the in-app feed, and delivery-channel linking —
 * {@code /api/me/notification-prefs}, {@code /api/me/notifications}, {@code /api/me/channels}.
 * All endpoints use the session token.
 */
public final class NotificationsResource extends Resource {

  NotificationsResource(Transport transport) {
    super(transport);
  }

  /** Read the caller's notification preferences. */
  public JsonNode getPrefs() {
    return transport.get("/api/me/notification-prefs", "session", null);
  }

  /**
   * Partially update preferences: absent fields stay unchanged, {@code quiet_hours: null}
   * clears quiet hours, unknown channel ids in {@code routing} are dropped server-side.
   */
  public JsonNode updatePrefs(Map<String, Object> params) {
    return transport.method("PATCH", "/api/me/notification-prefs", "session", params);
  }

  /**
   * The in-app feed, newest first — {@code {items, unread}}. {@code limit} 1–100 (default
   * 30); {@code before} is a {@code created_at} cursor for paging. Pass null to omit either.
   */
  public JsonNode list(Integer limit, String before) {
    Map<String, String> query = new HashMap<>();
    if (limit != null) {
      query.put("limit", String.valueOf(limit));
    }
    if (before != null) {
      query.put("before", before);
    }
    return transport.get("/api/me/notifications", "session", query.isEmpty() ? null : query);
  }

  /** Mark specific feed items read. */
  public JsonNode markRead(List<String> ids) {
    return transport.post("/api/me/notifications/read", "session", Collections.singletonMap("ids", ids));
  }

  /** Mark the whole feed read. */
  public JsonNode markAllRead() {
    return transport.post("/api/me/notifications/read", "session", Collections.singletonMap("all", true));
  }

  /** Linked delivery-channel identities plus linkable channel ids. */
  public JsonNode channels() {
    return transport.get("/api/me/channels", "session", null);
  }

  /**
   * Start linking a delivery channel; the verification code is delivered through the
   * channel itself and expires after 30 minutes. Bot channels accept an empty
   * {@code address} and answer with a one-time link code / deep link instead
   * ({@code status:'link_ready'}).
   */
  public JsonNode linkChannel(String channel, String address, String display) {
    Map<String, Object> body = new HashMap<>();
    body.put("channel", channel);
    body.put("address", address);
    if (display != null) {
      body.put("display", display);
    }
    return transport.post("/api/me/channels", "session", body);
  }

  /** Complete channel verification with the delivered code. */
  public JsonNode verifyChannel(String channel, String code) {
    Map<String, Object> body = new HashMap<>();
    body.put("channel", channel);
    body.put("code", code);
    return transport.post("/api/me/channels/verify", "session", body);
  }

  /** Unlink/revoke a channel identity. */
  public JsonNode unlinkChannel(String channel) {
    return transport.delete("/api/me/channels/" + enc(channel), "session");
  }
}
